### PR TITLE
[SPARK-24959][SQL] Speed up count() for JSON and CSV

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to skip invoking of the CSV/JSON parser per each line in the case if the required schema is empty. Added benchmarks for `count()` shows performance improvement up to **3.5 times**.

Before:

```
Count a dataset with 10 columns:      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)
--------------------------------------------------------------------------------------
JSON count()                               7676 / 7715          1.3         767.6
CSV count()                                3309 / 3363          3.0         330.9
``` 

After:

```
Count a dataset with 10 columns:      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)
--------------------------------------------------------------------------------------
JSON count()                               2104 / 2156          4.8         210.4
CSV count()                                2332 / 2386          4.3         233.2
```

## How was this patch tested?

It was tested by `CSVSuite` and `JSONSuite` as well as on added benchmarks.

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 1]
```diff
@@ -203,19 +203,11 @@ class UnivocityParser(
     }
   }
 
-  private val doParse = if (requiredSchema.nonEmpty) {
```
Comment (dmateusp) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: are the changes here https://github.com/apache/spark/pull/21909/files#diff-3a4dc120191f7052e5d98db11934bfb5R63 replacing the need for the `requiredSchema.nonEmpty` check ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala [hunk 2]
```diff
@@ -56,9 +57,14 @@ class FailureSafeParser[IN](
     }
   }
 
+  private val skipParsing = optimizeEmptySchema && schema.isEmpty
   def parse(input: IN): Iterator[InternalRow] = {
     try {
-      rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
+     if (skipParsing) {
+       Iterator.single(InternalRow.empty)
```
Comment (dmateusp) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: nit: `Iterator.empty`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 3]
```diff
@@ -2233,7 +2233,7 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
         .option("multiline", "true")
         .options(Map("encoding" -> "UTF-16BE"))
         .json(testFile(fileName))
-        .count()
+        .collect()
```
Comment (dmateusp) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: just curious why going from count() to collect() here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: It is not the same. If you return empty iterator, `count()` will always return `0`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: The test has to really touch JSON to detect encoding even without parsing. With this optimization `jackson` parser is not called at all in the case of `count()`. `collect()` guarantees that JSON parser will be invoked with wrong `encoding`.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 4]
```diff
@@ -450,7 +450,8 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
         input => rawParser.parse(input, createParser, UTF8String.fromString),
         parsedOptions.parseMode,
         schema,
-        parsedOptions.columnNameOfCorruptRecord)
+        parsedOptions.columnNameOfCorruptRecord,
+        optimizeEmptySchema = true)
```
[the remark this answers is not in the document]
Comment (dmateusp) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: No, no I'm just wondering since you made it a parameter that you can turn off and on, what would be the case to turn it off?

If there is none, shouldn't we just get rid of the parameter altogether ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala (hunk 2 above)
Comment (dmateusp) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: ohh yes my bad!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: > what would be the case to turn it off?

We can apply the optimization if we know in advance that one JSON object corresponds to one struct. In that case, we can return empty row if required schema (struct) is empty. If `multiLine` is enabled, there could be many structs per a JSON document. So, we cannot say in advance how many empty rows need to return without parsing.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmarks.scala [hunk 5]
```diff
@@ -119,8 +119,47 @@ object CSVBenchmarks {
...
+        .write
+        .csv(path.getAbsolutePath)
+
+      val ds = spark.read.schema(schema).csv(path.getAbsolutePath)
+
+      benchmark.addCase(s"Select $colsNum columns + count()", 3) { _ =>
+        ds.select("*").filter((_: Row) => true).count()
+      }
+      benchmark.addCase(s"Select 1 column + count()", 3) { _ =>
+        ds.select($"col1").filter((_: Row) => true).count()
```
Comment (felixcheung) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmarks.scala]: does this benchmark result vary if we select `col2` or `col10`?

**Comment by MaxGekk** [comment]
> does this benchmark result vary if we select col2 or col10?

@felixcheung Not so much. Here is the benchmark for CSV.
```
JJava HotSpot(TM) 64-Bit Server VM 1.8.0_172-b11 on Mac OS X 10.13.6
Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz

Count a dataset with 10 columns:         Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
col0 + count()                                9097 / 9167          1.1         909.7       1.0X
col2 + count()                                9294 / 9302          1.1         929.4       1.0X
col5 + count()                                9346 / 9394          1.1         934.6       1.0X
col7 + count()                                9227 / 9231          1.1         922.7       1.0X
col9 + count()                                9141 / 9233          1.1         914.1       1.0X
```

**Comment by felixcheung** [comment]
got it

**Comment by MaxGekk** [comment]
@HyukjinKwon @maropu Any objections to the PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala [hunk 6]
```diff
@@ -56,9 +57,14 @@ class FailureSafeParser[IN](
...
+     } else {
+       rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: If there are broken records the parser can't parse, this skipping won't detect them?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: Could you add a test case for counting both CSV and JSON source when the files having broken records? Any behavior change after this PR?

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: > ... when the files having broken records?

Syntactically broken or semantically (wrong types for example)?

> Any behavior change after this PR?

We have many tests in `CSVSuite` and `JSONSuite` for broken records. I have found behavior change in only one case: https://github.com/apache/spark/pull/21909/files#diff-fde14032b0e6ef8086461edf79a27c5dL2227 . This is due to `Jackson` parser touches a few first bytes in the input stream even if it is not called. `Jackson` checks encoding eagerly.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: both? 

If we introduce a behavior change, we need to document it in the migration guide and add a conf. Users can do the conf to revert back to the previous behaviors.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -1476,6 +1476,14 @@ object SQLConf {
         "are performed before any UNION, EXCEPT and MINUS operations.")
       .booleanConf
       .createWithDefault(false)
+
+  val BYPASS_PARSER_FOR_EMPTY_SCHEMA = buildConf("spark.sql.bypassParserForEmptySchema")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Let us get rid of this in the next release. Mark it as an internal and use the legacy scheme.

**Comment by gatorsmile** [comment]
Please document it in the migration guide.

**Comment by MaxGekk** [comment]
> Please document it in the migration guide.

@gatorsmile Updated by the notice:

  - Since Spark 2.4, text-based datasources like CSV and JSON don't parse input lines if the required schema pushed down to the datasources in read is empty. The schema can be empty in the case of count(), for example. To set `true` to `spark.sql.legacy.bypassParserForEmptySchema` restores the previous behavior when the underlying parser is always invoked even for the empty schema. This option will be removed in Spark 3.0.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 8]
```diff
@@ -2225,19 +2225,21 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
-    val exception = intercept[SparkException] {
-      spark.read.schema(schema)
-        .option("mode", "FAILFAST")
-        .option("multiline", "true")
-        .options(Map("encoding" -> "UTF-16BE"))
-        .json(testFile(fileName))
-        .count()
-    }
-    val errMsg = exception.getMessage
+    withSQLConf(SQLConf.BYPASS_PARSER_FOR_EMPTY_SCHEMA.key -> "false") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: How about CSV? Could you add the same one too?

Also, we need to add the verification logic when the conf is true. 
```
Seq(true, false).foreach { optimizeEmptySchema =>
  withSQLConf(SQLConf.BYPASS_PARSER_FOR_EMPTY_SCHEMA.key -> optimizeEmptySchema.toString) {
  ...
}
```

File: docs/sql-programming-guide.md [hunk 9]
```diff
@@ -1892,6 +1892,7 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
   - In version 2.3 and earlier, CSV rows are considered as malformed if at least one column value in the row is malformed. CSV parser dropped such rows in the DROPMALFORMED mode or outputs an error in the FAILFAST mode. Since Spark 2.4, CSV row is considered as malformed only when it contains malformed column values requested from CSV datasource, other values can be ignored. As an example, CSV file contains the "id,name" header and one row "1234". In Spark 2.4, selection of the id column consists of a row with one column value 1234 but in Spark 2.3 and earlier it is empty in the DROPMALFORMED mode. To restore the previous behavior, set `spark.sql.csv.parser.columnPruning.enabled` to `false`.
+  - Since Spark 2.4, text-based datasources like CSV and JSON don't parse input lines if the required schema pushed down to the datasources is empty. The schema can be empty in the case of count(), for example. To set `true` to `spark.sql.legacy.bypassParserForEmptySchema` restores the previous behavior when the underlying parser is always invoked even for the empty schema. This option will be removed in Spark 3.0.
```
Comment (cloud-fan) [inline docs/sql-programming-guide.md]: Let's highlight the behavior changing, i.e. malformed records won't be reported when doing `count()`. Does it also mean the result of `count()` can change if the json/csv files contain malformed records?

Comment (MaxGekk) [reply docs/sql-programming-guide.md]: For CSV, the result should be the same since we don't invoke the parser when CSV column pruning is enabled (by default). For JSON, result can be different like in the test which I changed in the PR. We call Jackson's method `nextToken()` even for empty schema, and besides of that the Jackson parser touches the input stream on instantiation. I will add a few tests for JSON as @gatorsmile asked.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 8 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: I will revert changes for the test because it enables multiLine mode for which the optimization is not applicable.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: I added the tests

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@HyukjinKwon @maropu Please, have a look at the PR.

File: docs/sql-programming-guide.md (hunk 9 above)
Comment (MaxGekk) [reply docs/sql-programming-guide.md]: > Does it also mean the result of count() can change if the json/csv files contain malformed records?

@cloud-fan Only the difference I have found so far is when a whole JSON file is invalid i.e it has wrong enconding, for example. In other cases `count()` returns the same.

**Comment by MaxGekk** [comment]
@cloud-fan @gatorsmile May I ask you to look at the PR again.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 10]
```diff
@@ -1492,6 +1492,15 @@ object SQLConf {
         "This usually speeds up commands that need to list many directories.")
       .booleanConf
       .createWithDefault(true)
+
+  val BYPASS_PARSER_FOR_EMPTY_SCHEMA =
+    buildConf("spark.sql.legacy.bypassParserForEmptySchema")
+      .doc("If required schema passed to a text datasource is empty, the parameter controls " +
+        "invocation of underlying parser. For example, if it is set to false, uniVocity parser " +
+        "is invoke by CSV datasource or Jackson parser by JSON datasource. By default, it is set " +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `invoke` -> `invoked`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 11]
```diff
@@ -2223,21 +2223,31 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
-        .options(Map("encoding" -> "UTF-16BE"))
-        .json(testFile(fileName))
-        .count()
+    def doCount(bypassParser: Boolean, multiLine: Boolean): Long = {
+      var result: Long = -1
+      withSQLConf(SQLConf.BYPASS_PARSER_FOR_EMPTY_SCHEMA.key -> bypassParser.toString) {
+        val fileName = "test-data/utf16LE.json"
+        val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+        result = spark.read.schema(schema)
+          .option("mode", "FAILFAST")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Does the mode matter? What happened if users use `DROPMALFORMED ` before this PR?

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: > Does the mode matter?

I just want to have an explicit error in the test instead of `0` for `count()` ( `DROPMALFORMED`), or full table of nulls or an exception (`PERMISSIVE`) since an exception is expected result.

> What happened if users use DROPMALFORMED before this PR?

It depends on `multiLine`. If it is `true`, behaviour before and after PR is the same since the optimization doesn't impact on the `multiLine` mode. For `multiLine` equals to `false`, after the PR the result is `5` (total number of lines), before the PR - `0` in the `DROPMALFORMED` mode.

We can enable this optimization for the `PERMISSIVE` mode only to exclude any deviation in outputs.

File: docs/sql-programming-guide.md [hunk 12]
```diff
@@ -1894,6 +1894,7 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
   - Since Spark 2.4, File listing for compute statistics is done in parallel by default. This can be disabled by setting `spark.sql.parallelFileListingInStatsComputation.enabled` to `False`.
   - Since Spark 2.4, Metadata files (e.g. Parquet summary files) and temporary files are not counted as data files when calculating table size during Statistics computation.
+  - Since Spark 2.4, text-based datasources like CSV and JSON don't parse input lines if the required schema pushed down to the datasources is empty. The schema can be empty in the case of the count() action. For example, Spark 2.3 and earlier versions failed on JSON files with invalid encoding but Spark 2.4 returns total number of lines in the file. To restore the previous behavior when the underlying parser is always invoked even for the empty schema, set `true` to `spark.sql.legacy.bypassParserForEmptySchema`. This option will be removed in Spark 3.0.
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: Is it right based on what you said https://github.com/apache/spark/pull/21909#discussion_r210704902?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 13]
```diff
@@ -1492,6 +1492,15 @@ object SQLConf {
         "This usually speeds up commands that need to list many directories.")
       .booleanConf
       .createWithDefault(true)
+
+  val BYPASS_PARSER_FOR_EMPTY_SCHEMA =
+    buildConf("spark.sql.legacy.bypassParserForEmptySchema")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: If no behavior change, do we still need this conf?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: It seems we don't need it anymore

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala [hunk 14]
```diff
@@ -223,7 +224,8 @@ object MultiLineJsonDataSource extends JsonDataSource {
       input => parser.parse[InputStream](input, streamParser, partitionedFileString),
       parser.options.parseMode,
       schema,
-      parser.options.columnNameOfCorruptRecord)
+      parser.options.columnNameOfCorruptRecord,
+      optimizeEmptySchema = false)
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala]: renamed

**Comment by gatorsmile** [comment]
LGTM.

Thanks for being patient to address all the comments! Merged to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala [hunk 15]
```diff
@@ -56,9 +58,15 @@ class FailureSafeParser[IN](
     }
   }
 
+  private val skipParsing = !isMultiLine && mode == PermissiveMode && schema.isEmpty
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: not a big deal but I would leave a comment to explain why it's permissive and non-miltiline only. I assume counts are known when it's actually parsed for multiline cases, and counts should be given in any case when the mode is permissive, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 16]
```diff
@@ -402,7 +403,7 @@ class JacksonParser(
         }
       }
     } catch {
-      case e @ (_: RuntimeException | _: JsonProcessingException) =>
+      case e @ (_: RuntimeException | _: JsonProcessingException | _: MalformedInputException) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: Is this change related, @MaxGekk? Let's don't add unrelated changes next time.
